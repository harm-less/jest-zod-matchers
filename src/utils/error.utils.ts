import {SafeParseError, SafeParseReturnType} from 'zod';
import {fromZodError} from 'zod-validation-error';

export function isSafeParseError(
	receivedSafeParseReturnType: SafeParseReturnType<unknown, unknown>
): receivedSafeParseReturnType is SafeParseError<unknown> {
	return receivedSafeParseReturnType.success === false;
}

export function handleSafeParseError(
	context: jest.MatcherContext,
	receivedResponse: SafeParseError<unknown>,
	{
		matcherHint
	}: {
		matcherHint: string;
	}
): jest.CustomMatcherResult {
	const receivedErrorMessage = getCleanZodError(receivedResponse).message;

	return {
		message: () =>
			matcherHint +
			'\n\n' +
			'Expected Zod schema input to be valid but a Zod error occured instead.' +
			'\n\n' +
			context.utils.BOLD_WEIGHT(`Zod error: ${receivedErrorMessage}`),
		pass: false
	};
}

export function getCleanZodError(receivedError: SafeParseError<unknown>) {
	return fromZodError(receivedError.error, {
		maxIssuesInMessage: 1,
		prefix: '',
		prefixSeparator: ''
	});
}
