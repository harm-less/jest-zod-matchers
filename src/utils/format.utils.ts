export function formatExpectedResult(obj: Record<string, unknown>): Record<string, unknown> {
	for (const propertyName in obj) {
		if (obj.hasOwnProperty(propertyName)) {
			const propertyValue = obj[propertyName];
			if (propertyValue instanceof Date) {
				obj[propertyName] = propertyValue.toISOString();
			} else if (typeof propertyValue === 'object' && propertyValue) {
				formatExpectedResult(propertyValue as Record<string, unknown>);
			}
		}
	}
	return obj;
}
