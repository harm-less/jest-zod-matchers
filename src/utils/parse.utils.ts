import {ParseParams, ZodRawShape, z} from 'zod';

export function parseObjectSchema(
	objectSchema: z.ZodObject<ZodRawShape>,
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	value: any,
	params?: Partial<ParseParams>
) {
	return objectSchema.safeParse(value, params);
}
