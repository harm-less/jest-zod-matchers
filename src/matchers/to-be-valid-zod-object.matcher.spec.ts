import {ZodEffects, ZodString, z} from 'zod';
import {parseObjectSchema} from '../utils/parse.utils';
import {toBeValidZodObject} from './to-be-valid-zod-object.matcher';

describe('"toBeValidZodObject" matcher', () => {
	let FooSchema: z.ZodObject<{
		foo: z.ZodString;
	}>;
	let TransformSchema: z.ZodObject<{
		foo: ZodEffects<ZodString, string, string>;
	}>;

	beforeEach(() => {
		expect.extend({toBeValidZodObject});

		FooSchema = z.object({
			foo: z.string()
		});
		TransformSchema = z.object({
			foo: z.string().transform((value) => `<${value}>`)
		});
	});

	test('parsed object to be valid', () => {
		expect(parseObjectSchema(FooSchema, {foo: 'bar'})).toBeValidZodObject();
	});

	test('parsed object with expected result to be valid', () => {
		expect(parseObjectSchema(FooSchema, {foo: 'bar'})).toBeValidZodObject<
			z.infer<typeof FooSchema>
		>({foo: 'bar'});
	});

	test('parsed object with transformed key to be valid', () => {
		expect(parseObjectSchema(TransformSchema, {foo: 'bar'})).toBeValidZodObject<
			z.infer<typeof FooSchema>
		>({foo: '<bar>'});
	});

	test('invalid parsed object not to be valid', () => {
		expect(parseObjectSchema(FooSchema, {foo: 'bar'})).not.toBeValidZodObject<
			z.infer<typeof FooSchema>
		>({foo: 'baz'});
	});

	describe('errors', () => {
		it('should throw error an error as the input is invalid', () => {
			expect(() =>
				expect(parseObjectSchema(FooSchema, {bar: 'foo'})).toBeValidZodObject()
			).toThrowErrorMatchingSnapshot();
		});

		test('invalid parsed object with expected result', () => {
			expect(() =>
				expect(parseObjectSchema(FooSchema, {foo: 'bar'})).toBeValidZodObject<
					z.infer<typeof FooSchema>
				>({foo: 'baz'})
			).toThrowErrorMatchingSnapshot();
		});

		test('invalid parsed object with transformed key', () => {
			expect(() =>
				expect(parseObjectSchema(TransformSchema, {foo: 'bar'})).toBeValidZodObject<
					z.infer<typeof FooSchema>
				>({foo: '<baz>'})
			).toThrowErrorMatchingSnapshot();
		});

		test('negated validation but without supplying zod output object', () => {
			expect(() =>
				expect(parseObjectSchema(FooSchema, {foo: 'bar'})).not.toBeValidZodObject<
					z.infer<typeof FooSchema>
				>()
			).toThrowErrorMatchingSnapshot();
		});

		test('zod object to not be valid', () => {
			expect(() =>
				expect(parseObjectSchema(FooSchema, {foo: 'bar'})).not.toBeValidZodObject<
					z.infer<typeof FooSchema>
				>({foo: 'bar'})
			).toThrowErrorMatchingSnapshot();
		});

		it('should throw an error if an expected property is not set in the received object', () => {
			expect(() =>
				expect(parseObjectSchema(FooSchema, {foo: 'bar'})).toBeValidZodObject({bar: 'bar'})
			).toThrowErrorMatchingSnapshot();
		});
	});
});
