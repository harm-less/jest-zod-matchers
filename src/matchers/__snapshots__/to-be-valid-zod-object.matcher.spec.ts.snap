// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"toBeValidZodObject" matcher errors invalid parsed object with expected result 1`] = `
"[2mexpect([22m[31mzodSafeParseReturnType[39m[2m).[22mtoBeValidZodObject[2m([22m[32mzodOutputObject[39m[2m) // Validate Zod object schema[22m

[32m- zodSafeParseData  - 1[39m
[31m+ zodOutputObject   + 1[39m

[2m  Object {[22m
[32m-   "foo": "baz",[39m
[31m+   "foo": "bar",[39m
[2m  }[22m"
`;

exports[`"toBeValidZodObject" matcher errors invalid parsed object with transformed key 1`] = `
"[2mexpect([22m[31mzodSafeParseReturnType[39m[2m).[22mtoBeValidZodObject[2m([22m[32mzodOutputObject[39m[2m) // Validate Zod object schema[22m

[32m- zodSafeParseData  - 1[39m
[31m+ zodOutputObject   + 1[39m

[2m  Object {[22m
[32m-   "foo": "<baz>",[39m
[31m+   "foo": "<bar>",[39m
[2m  }[22m"
`;

exports[`"toBeValidZodObject" matcher errors negated validation but without supplying zod output object 1`] = `
"[2mexpect([22m[31mzodSafeParseReturnType[39m[2m).[22mnot[2m.[22mtoBeValidZodObject[2m([22m[32mzodOutputObject[39m[2m) // Validate Zod object schema[22m

This test is negated with the ".not" keyword and the "zodOutputObject" object is undefined. Either make sure the received "zodSafeParseReturnType" is of type "SafeParseSuccess" or if you want to test Zod error messages use the "toThrowZodError" matcher instead of "toBeValidZodObject""
`;

exports[`"toBeValidZodObject" matcher errors should throw an error if an expected property is not set in the received object 1`] = `
"[2mexpect([22m[31mzodSafeParseReturnType[39m[2m).[22mtoBeValidZodObject[2m([22m[32mzodOutputObject[39m[2m) // Validate Zod object schema[22m

Property "bar" provided in expected result, but this property does not exist in the received Zod input object. Either remove key "bar" from your expected result, or make sure it's available in the received Zod input object"
`;

exports[`"toBeValidZodObject" matcher errors should throw error an error as the input is invalid 1`] = `
"[2mexpect([22m[31mzodSafeParseReturnType[39m[2m).[22mtoBeValidZodObject[2m([22m[32mzodOutputObject[39m[2m) // Validate Zod object schema[22m

Expected Zod schema input to be valid but a Zod error occured instead.

[1mZod error: Required at "foo"[22m"
`;

exports[`"toBeValidZodObject" matcher errors zod object to not be valid 1`] = `
"[2mexpect([22m[31mzodSafeParseReturnType[39m[2m).[22mnot[2m.[22mtoBeValidZodObject[2m([22m[32mzodOutputObject[39m[2m) // Validate Zod object schema[22m

zodSafeParseData: [32m{"foo": "bar"}[39m
zodOutputObject:  serializes to the same string"
`;
